### PR TITLE
dockerTools: Implement merging of image tarballs

### PR DESCRIPTION
**Pull request by considerate**
###### Motivation for this change

The `docker load` command supports loading tarballs that contain
multiple docker images with their respective image names and tags. This
enables distributing these images as a single file which simplifies the
release of software when an application requires multiple services to
run.

However, pkgs.dockerTools only create tarballs with a single docker
image and there exists is no mechanism in nixpkgs to combine the created
tarballs. This commit implements merging of tarballs in a way that is
compatible with `docker load`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Example

```console
$ nix repl .
Loading '.'...
Added 13742 variables.

nix-repl> bashImage = dockerTools.buildImage { name = "bash"; tag = "latest"; contents = [bashInteractive];}

nix-repl> redisImage = dockerTools.buildImage {name = "redis"; tag = "latest"; contents = [redis];}

nix-repl> :b dockerTools.mergeImages bashImage redisImage

this derivation produced the following outputs:
  out -> /nix/store/k92yyrcsq01q03849gzg4ppd5jsqj312-merge-docker-images
$ docker load < /nix/store/k92yyrcsq01q03849gzg4ppd5jsqj312-merge-docker-images
Loaded image: bash:latest
dcaeb884ff62: Loading layer  159.4MB/159.4MB
Loaded image: redis:latest
```

### CODE REVIEW

File: pkgs/build-support/docker/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:693]:
```suggestion
    mkdir a b image
```

File: pkgs/build-support/docker/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:699]:
```suggestion
    chmod -R +w a b
```

File: pkgs/build-support/docker/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:707]:
```suggestion
    # Move layers to output directory
    mv --no-clobber a/* image/
    mv --no-clobber b/* image/
```

File: pkgs/build-support/docker/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:711]:
```suggestion
    # Move merged repositories object and manifest list to output directory
    mv repositories image/repositories
    mv manifest.json image/manifest.json
```

**Review by roberth — CHANGES_REQUESTED** [review]
This is a great idea; thank you @considerate.

To reduce the amount of garbage in the Nix store, it would be better to implement this as a single derivation that processes a list of images. I know this derivation is prettier and can be used for the same goal, but the performance difference is rather significant because of the size.

Further improvement is possible if you back this function by a function `streamMergedImage`, similar to `streamLayeredImage`, but I don't think we need to require that. (Perhaps it would be considerate ;)  )

What we do need is a test for this function. You can add it to `nixos/tests/docker-tools.nix`. If you need a new example image, you can add it to `pkgs/build-support/docker/examples.nix`.
If I understand it correctly, you can also use this to produce a multi-arch image. You can test this by merging a cross-compiled version of the same image. This should also go in `nixos/tests/docker-tools.nix`, not `nixos/tests/docker-tools-cross.nix` because that file is for tests that need to _run_ the cross-compiled image as well.

**Comment by considerate** [comment]
@roberth, thank you very much for the nice comment and the review.

> To reduce the amount of garbage in the Nix store, it would be better to implement this as a single derivation that processes a list of images. I know this derivation is prettier and can be used for the same goal, but the performance difference is rather significant because of the size.

I agree, I didn't want to do that upfront in case it led to unnecessary complexity. Fortunately, it looks like that wasn't the case.

> Further improvement is possible if you back this function by a function `streamMergedImage`, similar to `streamLayeredImage`, but I don't think we need to require that. (Perhaps it would be considerate ;) )

Yes, this would indeed be nice (and considerate :P). I couldn't figure out what to do about the files that are created while
constructing the merged tarball though. Would a `mktemp -d` call be sufficient or is there some other more applicable way of handling these files when running outside of a derivation?

I tried to look at the Python code for the `streamLayeredImage` but as far as I can tell the only file written to was stdout. It's not impossible to do the merge of the `manifest.json` and `repositories.json` files in memory and stream the layer contents similar to what's done in the Python script but the complexity to achieve that is quite a bit greater than this bash script.

> What we do need is a test for this function. You can add it to `nixos/tests/docker-tools.nix`. If you need a new example image, you can add it to `pkgs/build-support/docker/examples.nix`.

I added a couple of basic tests that verify that the expected images are loaded.

> If I understand it correctly, you can also use this to produce a multi-arch image. You can test this by merging a cross-compiled version of the same image. This should also go in `nixos/tests/docker-tools.nix`, not `nixos/tests/docker-tools-cross.nix` because that file is for tests that need to _run_ the cross-compiled image as well.

I'm not sure if multi-arch images are possible. What the function does is that it adds multiple images into the same tarball. However, to have a multi-arch image you'd end up with one image per architecture and you'd have to make sure that the
tags for the images are different between architectures, otherwise you just get the last image in the list. What did you imagine the multi-arch image scenario looking like?

File: pkgs/build-support/docker/examples.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/examples.nix:416]:
```suggestion
  mergedBashAndRedis = pkgs.dockerTools.mergeImages [ bash redis ];

  mergedBashNoTagAndRedis = pkgs.dockerTools.mergeImages [ bashNoTag redis ];
```

File: pkgs/build-support/docker/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:693]:
```suggestion
    mkdir inputs image
```

**Comment by SuperSandro2000** [comment]
@roberth knows better if that works and collects garbage in the nix store. I just made the bash a bit shorter.

**Comment by considerate** [comment]
> @roberth knows better if that works and collects garbage in the nix store. I just made the bash a bit shorter.

Thank you for taking i a look at the code, @SuperSandro2000. I appreciate it!

File: pkgs/build-support/docker/default.nix
Comment (roberth) [inline pkgs/build-support/docker/default.nix:712]: > `--owner=0` etc

This changes the ownership info of the original images. This can be solved by running in fakeroot.
We should probably keep the mtimes from the original images as well.

**Comment by roberth** [comment]
> What did you imagine the multi-arch image scenario looking like?

The way I understood it was that the runtime would simply ignore the images with incompatible arch. That way a tarball with multiple images will have the desired behavior, despite having 'duplicate' images when only considering the name and tag.
I haven't fully read up on this though, so I may be wrong.

File: pkgs/build-support/docker/default.nix
Comment (roberth) [reply pkgs/build-support/docker/default.nix:712]: I thought for a minute `fakeroot` was already introduced in `buildLayeredImage`, but https://github.com/NixOS/nixpkgs/pull/116749 isn't merged yet. It may already be useful as an example of using fakeroot though.
For testing you could use `buildImage` with `runAsRoot` for the time being.

File: pkgs/build-support/docker/default.nix
Comment (considerate) [reply pkgs/build-support/docker/default.nix:704]: I did this initially but opted for `cp` since `mv` requires changing the permissions of the source images.

**Comment by considerate** [comment]
@roberth, thank you for the review!

> > What did you imagine the multi-arch image scenario looking like?
> 
> The way I understood it was that the runtime would simply ignore the images with incompatible arch. That way a tarball with multiple images will have the desired behavior, despite having 'duplicate' images when only considering the name and tag.
> I haven't fully read up on this though, so I may be wrong.

I looked into multi-arch images a bit and it seems like `docker` provides an experimental `buildx` command which indeed seems to have that behavior. Unfortunately, I think the format is different to the tarball format expected by `docker load`. You might be correct in that it just works under the hood by the load command, if that's the case I'll add a test case that verifies this. If not, it would still be nice to look into what would be required to support this new format as well, though.

File: pkgs/build-support/docker/default.nix
Comment (considerate) [reply pkgs/build-support/docker/default.nix:712]: I don't really understand all the implications of permissions and the `--xform` flag. I simply copied how it was done in the `buildImage` command. I'll try to do this in a `fakeroot` and keep the modification times from the source images.

**Comment by considerate** [comment]
@roberth I created a tarball consisting of two images for different architectures (musl64, aarch64), each with the image and tag `hello-cross:latest`. This is the output I get after loading the tarball and running the `hello-cross:latest` image:

```console
$ docker load < ./result
The image hello-cross:latest already exists, renaming the old one with ID sha256:60289b88add7213b27bf7434c55819769eb96ac23265094894df9d5a7224af3e to empty string
Loaded image: hello-cross:latest
The image hello-cross:latest already exists, renaming the old one with ID sha256:b3bda3d5851a534c8ff91b674b92ac82d1d65584e109ac3d272ee66de773db51 to empty string
Loaded image: hello-cross:latest
```

```console
$ docker run hello-cross:latest hello
WARNING: The requested image's platform (linux/arm64) does not match the detected host platform (linux/amd64) and no specific platform was requested
standard_init_linux.go:219: exec user process caused: no such file or directory
```
Hence, the multi-arch image didn't work  and it picked up the arm image despite the musl64 image being runnable on my system.

**Comment by roberth** [comment]
Could you rebase and add a test to confirm that ownership is preserved? I think you need to add fakeroot to the `tar -x` invocation as well, because otherwise the original ownership info isn't recorded. I don't know if it's ok to invoke fakeroot twice. If not, a large portion of the script needs to be in fakeroot. The test will point this out.

**Comment by roberth** [comment]
You can reuse this example https://github.com/NixOS/nixpkgs/pull/116749/files#diff-730cb9cd318fb9539b54b5501b42359537f7e3280864257df2e1f73c58848cae

**Comment by considerate** [comment]
@roberth I rebased and tested that the `/home/jane` directory had the correct owner. However, then I tested without `fakeroot` and that also worked since the permissions of the filesystem content comes from the `layer.tar` files of the individual layers and not from the final `image.tar` tarball. Hence, I'm not sure whether I should just revert to not using `fakeroot` or if there is a need to be careful about permissions in the top-level tarball.

**Comment by considerate** [comment]
Also I noticed that the fakeroot layered image didn't have a `repositories` file so I updated the `mergeImages` script to only merge repositories from tarballs that contain that file (likewise for the manifest.json file).

**Comment by roberth** [comment]
Sorry, I was confused by the nesting of tarballs. Your first approach was correct all along. Please revert and make sure to produce a deterministic tarball. The [original line](https://github.com/NixOS/nixpkgs/pull/118018#discussion_r605912436) looks ok.

**Comment by considerate** [comment]
> Sorry, I was confused by the nesting of tarballs. Your first approach was correct all along. Please revert and make sure to produce a deterministic tarball. The [original line](https://github.com/NixOS/nixpkgs/pull/118018#discussion_r605912436) looks ok.

No problem, it wasn't clear to me before testing it out either. I have now reverted the change and tested that it still works correctly.

**Comment by considerate** [comment]
I looked into the multi-arch images a bit more, I don't know for sure yet but it looks like it requires image [manifest version 2](https://docs.docker.com/registry/spec/manifest-v2-2/) but the docker images generated in nixpkgs follow the [manifest version 1.2 format](https://github.com/moby/moby/blob/master/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format).

As far as I can tell the version 2 manifest would replace the `<hash>.json` files referenced by the `"Config"` field in the top-level `manifest.json`. Nonetheless, I think it'd be better to implement that in a different PR, especially since `docker save` outputs the version 1.2 format on my system:
```console
$ docker --version
Docker version 20.10.2, build v20.10.2
```

**Comment by considerate** [comment]
@roberth Can I get a second review of the code now that the original points have been addressed?

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
I agree that multi-arch is out of scope.

**Comment by considerate** [comment]
@roberth Thank you for merging this!

**Comment by roberth** [comment]
> especially since `docker save` outputs the version 1.2 format on my system:

"You can rebuild the image by updating the FROM statement in your Dockerfile. If your image manifest is out-of-date, there is a chance the image pulled from your FROM statement in your Dockerfile is also out-of-date."

https://docs.docker.com/registry/spec/deprecated-schema-v1/#update-from-statement

**Pull request merged**
